lstchain/scripts/lstchain_r0_to_r0g.py

```diff
@@ -0,0 +1,206 @@
+
+parser.add_argument('-o', '--output-dir', dest='output_dir',
+                    type=str, default='./',
+                    help='Output directory')
+
+# Range of waveform to be checked (for gain selection & heuristic FF
+# identification)
+SAMPLE_START = 3
+SAMPLE_END = 39
+
+def main():
+    args = parser.parse_args()
+
+    # Level of high gain (HG) required for switching to low gain (LG)
+    THRESHOLD = 3900
+
+    input_file = args.input_file
+    output_dir = args.output_dir
+
+    # First identify properly interleaved pedestals (also in case there are
+    # ucts jumps) and FF events (heuristically):
+    event_id, event_type = get_event_types(input_file)
+
+    event_id = np.array(event_id)
+    event_type_val = np.array([x.value for x in event_type])
+
+    logging.info('Identified event types and number of events:')
+    for j in np.unique(event_type_val):
+        logging.info(f'{EventType(j)}:, {np.sum(event_type_val == j)}\n')
+
+    # Now loop over the files (4 streams) again to perform the actual gain
+    # selection:
+    k = input_file.find('/LST-1.')
+    input_stream_names = [input_file[:k+7]+str(id_stream+1)+input_file[k+8:]
+                          for id_stream in range(4)]
+    output_stream_names = [output_dir + name[k:] for name in input_stream_names]
+
+    input_streams = []
+    for name in input_stream_names:
+        input_streams.append(protozfits.File(name, pure_protobuf=True))
+
+    try:
+        num_pixels = input_streams[0].CameraConfiguration[0].num_pixels
+        num_samples = input_streams[0].CameraConfiguration[
+            0].num_samples_nominal
+    except:
+        logging.error('CameraConfiguration not found! Is this R1v1 data?')
+        exit(1)
+
+    with ExitStack() as stack:
+        for i, name in enumerate(output_stream_names):
+
+            n_tiles = input_streams[i].Events.protobuf_i_fits.header[
+                'NAXIS2'].value
+            rows_per_tile = input_streams[i].Events.protobuf_i_fits.header[
+                'ZTILELEN'].value
+
+            stream = stack.enter_context(protozfits.ProtobufZOFits(
+                    n_tiles=n_tiles,
+                    rows_per_tile=rows_per_tile,
+                    compression_block_size_kb=64*1024,
+                    defaul_compression="lst"))
+            stream.open(name)
+
+            stream.move_to_new_table("CameraConfiguration")
+            stream.write_message(input_streams[i].CameraConfiguration[0])
+            stream.move_to_new_table("DataStream")
+            stream.write_message(input_streams[i].DataStream[0])
+
+            stream.move_to_new_table("Events")
+
+            for event in input_streams[i].Events:
+                # skip corrupted events:
+                if event.event_id == 0:
+                    continue
+                # Get the event type, from the first loop over the files:
+                evtype = event_type_val[event_id==event.event_id][0]
```

Comment:
This is very inefficient to for each event build a new mask and check the full array. Can use a dict or a table with an index to lookup by the event id instead of doing a full search for each event.